### PR TITLE
Replace Guava cache with Caffeine

### PR DESCRIPTION
**Pull request by Vlatombe**
* Replace guava cache with Caffeine

- [X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [X] Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Comment by jglick** [comment]
Relates to https://github.com/jenkinsci/workflow-cps-plugin/pull/492.

**Review by jglick — COMMENTED** [review]
Caffeine switch is certainly good. I am not convinced we want to switch from soft to weak references; seems like it would wind up causing increased disk I/O (and some CPU) under typical load. The root cause of the large heap allocations and long GC pauses mentioned in the description turned out to be incomplete truncation behavior in `ArgumentsActionImpl`, which is already being corrected in a separate PR, so this PR would penalize instances with more normal metadata (and just a lot of builds and steps).

File: src/main/java/org/jenkinsci/plugins/workflow/support/storage/SimpleXStreamFlowNodeStorage.java [hunk 1]
```diff
@@ -71,11 +69,11 @@
...
     private final File dir;
     private final FlowExecution exec;
-    private final LoadingCache<String,FlowNode> nodeCache = CacheBuilder.newBuilder().softValues().build(new CacheLoader<String,FlowNode>() {
-        @Override public FlowNode load(String key) throws Exception {
-            return SimpleXStreamFlowNodeStorage.this.load(key).node;
-        }
-    });
+
+    private final LoadingCache<String,FlowNode> nodeCache = Caffeine.newBuilder()
+            .weakKeys()
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/support/storage/SimpleXStreamFlowNodeStorage.java:74]: Also, why weak _keys_? These should be small `String`s, currently always numeric like `123`. We could perhaps `intern` them, but beyond that…?

**Comment by Vlatombe** [comment]
I agree with you, the cache problem here is a consequence of the incomplete truncation. Better progress on the other, and possibly keep the switch to Caffeine. Soft vs. weak references could be reconsidered later.

**Comment by car-roll** [comment]
LGTM  if we just do the switch to caffeine

**Review by jglick — APPROVED** [review]

**Review by car-roll — APPROVED** [review]

**Pull request merged**